src/RequestHeaders.cs

```diff
@@ -0,0 +1,129 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Kiota.Abstractions;
+
+/// <summary>Represents a collection of request headers.</summary>
+public class RequestHeaders : IDictionary<string,IEnumerable<string>> {
+    internal Dictionary<string, HashSet<string>> _headers = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
+    /// <summary>
+    /// Adds values to the header with the specified name.
+    /// </summary>
+    /// <param name="headerName">The name of the header to add values to.</param>
+    /// <param name="headerValues">The values to add to the header.</param>
+    public void Add(string headerName, params string[] headerValues) {
+        if(string.IsNullOrEmpty(headerName))
+            throw new ArgumentNullException(nameof(headerName));
+        if(headerValues == null)
+            throw new ArgumentNullException(nameof(headerValues));
+        if(!headerValues.Any())
+            return;
+        if(_headers.TryGetValue(headerName, out var values))
+            foreach(var headerValue in headerValues)
+                values.Add(headerValue);
+        else
+            _headers.Add(headerName, new HashSet<string>(headerValues));
+    }
+    /// <inheritdoc/>
+    public ICollection<string> Keys => _headers.Keys;
+    /// <inheritdoc/>
+    public ICollection<IEnumerable<string>> Values => _headers.Values.Cast<IEnumerable<string>>().ToList();
```

Comment:
```suggestion
    public ICollection<IEnumerable<string>> Values => (ICollection<IEnumerable<string>>)_headers.Values.Cast<IEnumerable<string>>();;
```
To avoid the performance hit involved creating a new List every time this is called and iterating through the items to copy them over, maybe we could do a direct cast as we already know that Values already implements [`ICollection<T>`](https://github.com/microsoft/referencesource/blob/dae14279dd0672adead5de00ac8f117dcf74c184/mscorlib/system/collections/generic/dictionary.cs#L988)

Reply:
I'm not sure about this one. I can't cast the values collection directly because if covariance as far as I know. And casting after the cast method call really depends on the internals of LINQ, which might be a faster but riskier approach.

Reply:
True. Better the safer approach for now.